engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnJobBehavior.java

```diff
@@ -171,20 +172,15 @@ public void cancelJob(final BpmnElementContext context) {
     final var elementInstance = stateBehavior.getElementInstance(context);
     final long jobKey = elementInstance.getJobKey();
     if (jobKey > 0) {
-      writeJobCancelCommand(jobKey);
+      final State state = jobState.getState(jobKey);
+      if (CANCELABLE_STATES.contains(state)) {
+        final JobRecord job = jobState.getJob(jobKey);
+        stateWriter.appendFollowUpEvent(jobKey, JobIntent.CANCELED, job);
```

Comment:
:question:/:x: Can you talk about why you write a `JobIntent.CANCELLED` now instead of using the old `JobIntent.CANCEL` command? (which in turn writes the `JobIntent.CANCELLED` event)

The difference currently is that `jobMetrics.jobCanceled(job.getType());` is not called anymore (which happens in `JobCancelProcessor`. And as soon as we add more logic to `JobCancelProcessor` it will be even more complicated to keep the logic in sync if this call site circumvents the command.

Reply:
The idea is that writing a command postpones canceling the job to when that command is processed. When the Incident Resolved event is applied shortly after the above code, it looks for a related job and then makes it activatable again. If the command is written, the job is not yet canceled/deleted. So the incident resolved event applier makes the job activatable again. Of course, a short bit later the job is canceled, but there is some period where the job is available to be activated.

IMO there is too much logic in this IncidentResolved event applier. However, I'd rather avoid making a change to the event appliers (remember that events should always lead to the same state, even if that means keeping the result of a bug). We could avoid that by building support for version-dependent event appliers, but we don't have that yet.

Instead, we can simply immediately cancel the job by writing the job canceled event, which deletes the job. The change is small enough to be backported. Does that make sense?

👍 You make a good point about the job metrics, which I completely missed. I'll make sure to add the metric here as well. 🤔 Why is there not a test for this?

Reply:
Could you also look at whether `JobCancelProcessor` could get a dependency to `BpmnJobBehavior` ? 

Then we might be able to avoid code duplication if we split up the `cancelJob()` method in `BpmnJobBehavior` in `cancelByWorker()` (without resolving incident) and `cancelBycontainer()` (with resolving incident)

Reply:
We've discussed this F2F, but here's a recap. The cancel job command was only written by the engine, it is not available from the gRPC API. Since we'll no longer be writing the command we will deprecate it (this follows in another PR). We'll have some duplicated code between the JobCancelProcessor and the BpmnJobBehavior, but we'll document the duplication in both places.